notebooks/working/ho_deconfounder_on_sim_data.py

```diff
@@ -0,0 +1,811 @@
+
+# +
+pval_mode = []
+for mode in range(len(holdouts_req)):
+    pvals = np.array([np.mean(rep_ll_per_zi_per_mode[mode][:,i] < obs_ll_per_zi_per_mode[mode][i]) 
+                      for i in range(holdouts_req[mode].shape[0])])
+    holdout_subjects = np.unique(rows_df[mode])
+    overall_pval = np.mean(pvals[holdout_subjects])
+    pval_mode.append(overall_pval)
+#     print("Predictive check p-values", overall_pval)
+
+pval_mode
+# -
+
+len(rows_df[6])
+
+# +
+holdout_subjects_0 = np.unique(rows_df[0])
+
+subject_no = npr.choice(holdout_subjects_0) 
+sns.kdeplot(rep_ll_per_zi_per_mode[0][:,subject_no]).set_title("Predictive check for subject "+str(subject_no))
+plt.axvline(x=obs_ll_per_zi_per_mode[0][subject_no], linestyle='--')
+# -
+
+# ### Adding confounders to original DF
+
+# +
+for i in data['mode_id'].unique():
+    
+#     print(len(data.loc[data['mode_id']==i, col_name]), len(confounder_vectors[int(i-1)][2]) )
+    
+    col_name = 'confounder_for_mode_' + str(int(i))
+    data.loc[data['mode_id']==i, col_name] = confounder_vectors[int(i-1)][2]
+    data[col_name] = data[col_name].fillna(0)
+    
+data['confounder_all'] = data[['confounder_for_mode_1','confounder_for_mode_2','confounder_for_mode_3',
+                              'confounder_for_mode_4', 'confounder_for_mode_5', 'confounder_for_mode_6',
+                              'confounder_for_mode_7', 'confounder_for_mode_8']].sum(axis=1)
+# -
+
+data['confounder_all_2'] = confounders_using_all[2]
+
+# ## Estimate original MNL
+
+# data
+
+
+# +
+# Estimate the basic MNL model, using the hessian and newton-conjugate gradient
+mnl_model = cm.create_choice_model(data=data,
+                                   alt_id_col="mode_id",
+                                   obs_id_col="observation_id",
+                                   choice_col="sim_choice",
+                                   specification=mnl_specification,
+                                   model_type="MNL",
+                                   names=mnl_names)
+
+num_vars = len(reduce(lambda x, y: x + y, mnl_names.values()))
+# Note newton-cg used to ensure convergence to a point where gradient 
+# is essentially zero for all dimensions. 
+mnl_model.fit_mle(np.zeros(num_vars),
+                  method="BFGS")
+
+# Look at the estimation results
+mnl_model.get_statsmodels_summary()
+# -
+
+# ## Estimate non-causal MNL -- omit travel distance
+
+# +
+# Create my specification and variable names for the basic MNL model
+# NOTE: - Keys should be variables within the long format dataframe.
+#         The sole exception to this is the "intercept" key.
+#       - For the specification dictionary, the values should be lists
+#         or lists of lists. Within a list, or within the inner-most
+#         list should be the alternative ID's of the alternative whose
+#         utility specification the explanatory variable is entering.
+
+mnl_specification_noncausal = OrderedDict()
+mnl_names_noncausal = OrderedDict()
```

Comment:
The following might be much clearer for readers than recreating the name dictionary from scratch.
```suggestion
mnl_names_noncausal = deepcopy(mnl_names)
del mnl_names_noncausal['total_travel_distance']
```

Note you may need to do `from copy import deepcopy`.